internal/persistence/sql/traverser.go

```diff
@@ -0,0 +1,189 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package sql
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/gobuffalo/pop/v6"
+	"github.com/gofrs/uuid"
+	"github.com/ory/x/otelx"
+	"github.com/ory/x/sqlcon"
+	"github.com/pkg/errors"
+
+	"github.com/ory/keto/internal/namespace/ast"
+	"github.com/ory/keto/internal/relationtuple"
+	"github.com/ory/keto/ketoapi"
+)
+
+type (
+	Traverser struct {
+		conn *pop.Connection
+		d    dependencies
+		nid  uuid.UUID
+		p    *Persister
+	}
+
+	subjectExpandedRelationTupleRow struct {
+		RelationTuple
+
+		Found bool `db:"found"`
+	}
+
+	rewriteRelationTupleRow struct {
+		RelationTuple
+		Traversal relationtuple.Traversal `db:"traversal"`
+	}
+)
+
+func whereSubject(sub relationtuple.Subject) (sqlFragment string, args []any, err error) {
+	switch s := sub.(type) {
+	case *relationtuple.SubjectID:
+		sqlFragment = "subject_id = ? AND subject_set_namespace IS NULL AND subject_set_object IS NULL AND subject_set_relation IS NULL"
+		args = []any{s.ID}
+
+	case *relationtuple.SubjectSet:
+		sqlFragment = "subject_id IS NULL AND subject_set_namespace = ? AND subject_set_object = ? AND subject_set_relation = ?"
+		args = []any{s.Namespace, s.Object, s.Relation}
+
+	case nil:
+		return "", nil, errors.WithStack(ketoapi.ErrNilSubject)
+	}
+	return sqlFragment, args, nil
+}
+
+func (t *Traverser) TraverseSubjectSetExpansion(ctx context.Context, start *relationtuple.RelationTuple) (res []*relationtuple.TraversalResult, err error) {
+	ctx, span := t.d.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.TraverseSubjectSetExpansion")
+	defer otelx.End(span, &err)
+
+	targetSubjectSQL, targetSubjectArgs, err := whereSubject(start.Subject)
+	if err != nil {
+		return nil, err
+	}
+
+	var rows []*subjectExpandedRelationTupleRow
+	err = t.conn.RawQuery(fmt.Sprintf(`
+SELECT current.subject_set_namespace AS namespace,
+       current.subject_set_object AS object,
+       current.subject_set_relation AS relation,
+       EXISTS(
+           SELECT 1 FROM keto_relation_tuples
+           WHERE nid = current.nid AND
+                 namespace = current.subject_set_namespace AND
+                 object = current.subject_set_object AND
+                 relation = current.subject_set_relation AND
+                 %s  -- subject where clause
+       ) AS found
+FROM keto_relation_tuples AS current
+WHERE current.nid = ? AND
+      current.namespace = ? AND
+      current.object = ? AND
+      current.relation = ? AND
+      current.subject_id IS NULL
+`, targetSubjectSQL),
+		append(targetSubjectArgs, t.p.NetworkID(ctx), start.Namespace, start.Object, start.Relation)...,
+	).All(&rows)
```

Comment:
Yes that is true. But instead of pagination, how do you feel about just truncating it? In practice, can we handle fan-out on subject-sets the same set as too-deeply-nested hierarchies?

Reply:
The problem will be that we have no deterministic way to do that. A depth limit always returns the same result, but if we just randomly truncate the width this could be a real issue. Also, the application of Keto will be very limited when we start to impose this kind of limit, as I expect a lot of wide trees (groups with many members, documents that are shared with many people, many documents in one directory, ....). So IMO it is not acceptable to just truncate the result here.
The performance with very wide graphs can only really be improved with a "shortcut index" (Leopard in the paper).